tests/pthread/test_pthread_detach.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2019 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <thread>
+#include <math.h>
+#include <emscripten.h>
+#include <assert.h>
+
+#ifndef REPORT_RESULT
+#include <iostream>
+#endif
+
+extern "C" {
+//Create a thread that does some work
+void EMSCRIPTEN_KEEPALIVE spawn_a_thread() {
+	std::thread( [] {
+		double d=0;
+		for (int i=0; i<10000000; i++)			//simulate work
+			d += (i%2 ? sqrt((int)(rand())) : (-1)*sqrt((int)(rand())));
+#ifndef REPORT_RESULT
+		std::cout << d << std::endl;
+#endif
+	} ).detach();
+}
+
+
+//Check that the number of workers is less than the number of spawned threads.
+void EMSCRIPTEN_KEEPALIVE count_threads(int num_threads_spawned) {
+	int num_threads = EM_ASM_INT({
+		return PThread.runningWorkers.length + PThread.unusedWorkerPool.length;
+	});
+
+#ifdef REPORT_RESULT
+	if (num_threads < num_threads_spawned)
+		REPORT_RESULT(0);
+	else
+		REPORT_RESULT(num_threads);
+#else
+	assert(num_threads < num_threads_spawned);
+	std::cout << num_threads << std::endl;
+#endif
+}
+}
+
+//Spawn a detached thread every 0.5s. After 10s Check that the number of workers are less than the number of spawned threads
+int main(int argc, char** argv) {
+	EM_ASM(
+		const i_max = 20;
+		for (let i=0; i<i_max; i++) {
+			setTimeout(() => { _spawn_a_thread(); }, i*500);
+		}
+		
+		setTimeout(() => { _count_threads(i_max); }, i_max*500);
```

Comment:
This is a clever way to test, though can we make this test much shorter? A 10 second long test is quite long in comparison to most tests in the suite, so these kind of things add up the overall suite runtime. Perhaps this can be tested e.g. by spawning one thread detached, then exiting it, and then when that is confirmed to have exited (e.g. via an atomic var), wait for 500 msecs for good measure, and then run a second thread, which should be run in the context of the same worker that ran the first thread. That way the test should finish in less than a second, and if the test needs to be duplicated in multiple modes in the future (PROXY_TO_PTHREAD & asm.js vs wasm & different opts modes are common different cases that often are needed to test), that will not multiply the overall test suite run by that much.

Reply:
Redid this.
Yeah you were right, a 10s test is far too long. It should now take about 1/2s. I didn't use an atomic, because I figured there was going to be some time between a thread going out of scope and being returned to the pool.
It's not completely race free, but I'm performing 5 100ms checks on the thread status, so if even one thread hasn't returned something has probably gone wrong.